Proposal: Use post previews on the homepage

mkasberg opened the pull request:
I consider this an idea - a starting point for a conversation.
Happy to merge this as-is, or work on some changes/improvements, or
discard it if you don't like the idea.

Currently, the homepage is a list of full-length posts, paginated with 7
per page. This requires a lot of scrolling, and it seems like a cleaner
design would be to show a preview of each post in the list on the
homepage, but require clicking into the post itself to continue reading.
The homepage looks a little cleaner, and you're more immersed in the
post itself (no other posts, and the comments are expanded) when you
read one.

There are probably a few different strategies for generating the
preview. The one I'm using here takes the first 40 words. It's somewhat
important to render HTML and markdown before truncating - otherwise
posts that have HTML or markdown content near the top might have
non-displayed words counting against them.

![image](https://user-images.githubusercontent.com/12214617/96819351-a2cedc80-13e0-11eb-9540-e99236728cc3.png)

mkasberg committed:
Use post previews on the homepage

Currently, the homepage is a list of full-length posts, paginated with 7
per page. This requires a lot of scrolling, and it seems like a cleaner
design would be to show a preview of each post in the list on the
homepage, but require clicking into the post itself to continue reading.
The homepage looks a little cleaner, and you're more immersed in the
post itself (no other posts, and the comments are expanded) when you
read one.

There are probably a few different strategies for generating the
preview. The one I'm using here takes the first 40 words. It's somewhat
important to render HTML and markdown before truncating - otherwise
posts that have HTML or markdown content near the top might have
non-displayed words counting against them.

geerlingguy commented on _config.yml:
```diff
@@ -11,6 +11,7 @@ github_username:  opensourcecatholic
 mailinglist_name: open-source-catholic
 permalink: /:year/:title
 paginate: 7
+post_preview_words: 42
```
I'd say maybe double this, just so its easier to get a good idea of the contents on the first glance.

geerlingguy commented on _includes/post_preview.html:
```diff
@@ -0,0 +1,24 @@
+</h2>
+
+<div class="post-content" itemprop="articleBody">
+  {{ post.content | markdownify | strip_html | truncatewords: site.post_preview_words }}
```
Is there any way to truncate/strip without stripping everything? E.g. having paragraphs to break up the first couple sections of text may be helpful to scan an article (especially in combination with an expanded value for `post_preview_words`).

mkasberg replied:
Not completely sure, but I bet we can do better than what I have here. I'll spend a little time digging into this and see what I can figure out.

mkasberg replied:
Found a simple plugin that does the job pretty well

![image](https://user-images.githubusercontent.com/12214617/97522108-99a3b980-1964-11eb-9cfd-bebbd892ffeb.png)

mkasberg committed:
Improve post previews on homepage

In the previous commit, I added post previews to the homepage. Here, I'm
improving that experience. The first iteration of this functionality
stripped HTML and rendered the post preview as plain text. This was
required because it did simple text truncation, and it's problematic to
truncate in the middle of an HTML tag.

I'm now improving on that by using a plugin to do smart_truncate. This
plugin is aware of HTML markup and will ensure the HTML tag is closed
appropriately when we truncate. This means HTML will render correctly in
the preview even when we truncate! This is possible due to the plugin I
found referenced here:
https://yingtongli.me/blog/2016/07/12/html-truncate.html. We're allowed
to use it because it's open source under the GNU Affero GPL. Small
bummer that it's distributed as a ruby file rather than a gem and it
doesn't come with unit tests, but the code's short enough I don't view
this as a big concern.

I also double the number of words included in the preview to provide a
little more context to each post.

geerlingguy commented:
Ooh, looks interesting; do you know if GitHub Pages will load this custom plugin properly? I know Jekyll can locally but I don't know about GH Pages in particular.

mkasberg commented:
Good question. I think there are a couple different ways to deploy github pages these days, each with their own restrictions. I was under the impression we _can_ add plugins like this because I don't see a [github-pages](https://github.com/github/pages-gem) dependency in our Gemfile, but let me see what I can find out about our build process.

mkasberg commented:
OK, I was incorrect, we can't use that plugin, at least with our current deployment process.

I think our current setup is actually mildly broken, though probably not in a way that's actually noticeable yet. Our Gemfile specifies Jekyll 4.1.

https://github.com/opensourcecatholic/opensourcecatholic.github.io/blob/a674f3672d9c05f8448065704b444db296c16db0/Gemfile#L3

But when we deploy to GitHub pages, the [build process](https://pages.github.com/versions/) actually runs Jekyll 3.9. This means Jekyll 4 stuff would work locally (with `jekyll serve`) but fail when deployed. (We must not be using any Jekyll 4 stuff.)

One option is to keep our current deploy process. If we go that route, I think I'd want to open a PR that adds the [github-pages](https://github.com/github/pages-gem) gem as a dependency in our Gemfile, which should help maintain a consistent local env with what we're building for prod.

On the other hand, we could switch our deploy process to use [GitHub Actions](https://jekyllrb.com/docs/continuous-integration/github-actions/). This is the cool new way to do things. It allows you to use Jekyll 4.x, and also allows you to use any plugins you want. I think I'd recommend this approach. It's getting late here tonight, but I can take a closer look at this in the near future. Happy to talk more here or on Slack if you want to chat!

geerlingguy commented:
We can fix any issues that result in follow-ups. Thanks for the work on this, it'll make it a lot nicer looking blog!
